### ISSUE TITLE
GPU support with cupy

### ISSUE THREAD

**Issue by VolkerH** [david-hoffman/pyDecon#5]
Hi, 

I had forked this repo a while ago to play around with the code and see whether I can get the accelerated code run with gputools.
I got halfway there but then had to work on other stuff. Now, a few months later and just having received a new machine with CUDA GPU I tried again porting your core RL algorithm to cupy. This is straightforward and gives a significant speed-up.

As I was pushing this to my fork I noticed that you have done some recent work on the repo and are also thinking of packaging it. 
Not sure whether you have time to integrate cupy support ?

Here is a code snippet based on your code that can run both with and without gpu (depending on the thruthvlaue of the GPU flag at the top): https://github.com/VolkerH/pyDecon/blob/cupy/decon_cupy.py

This is stripped down to a single file, with just the "matlab" inner and the option for first-order Andrews/Biggs acceleration.

**Comment by david-hoffman** [comment]
I would love to have GPU support! I tried a long time ago with Numba and failed totally (the arrays were always transferred back to the CPU after the GPU operation). Do you have any benchmarks comparing the two? I'm also curious as to whether you get a substantial speed up using the direct convolution method, I could imagine for small kernels this would better utilize the multiple cores on a GPU than an FFT would, but that depends on how optimized the cuFFT routines are.

**Comment by david-hoffman** [comment]
And, of course, issue #1 might be relevant here 😅

**Comment by VolkerH** [comment]
Hi,
thanks for the reply.

Regarding issue #1 : I am aware of (and have been using) other GPU accelerated python deconvolution packages, including flowdec and pycudadecon. They are good and give great results. The only issue I have is that they are not as easy and pythonic to use as a basis for modifications/experiments with as puycudadecon wraps a C++/Cuda library and flowdec builds a static tensorflow compute graph that is then executed. Your plain python/numpy code is much easier in that regard. You can just debug at any stage during the iteration and look at intermediate results without leaving the python ecosystem.

Regarding benchmarking: 

I haven't done well-controlled benchmarks (there may be caching effects etc.), but to give some indication I have some notes in my fork:
https://github.com/VolkerH/pyDecon/issues/1

The benchmarks there are on the DecovolutionLab2 drosophila embryo sample volume showing the iterations per second for GPU and CPU. In each screenshot the upper progress bar is using Andrew-Biggs acceleration and the lower one is without.

In the case without Andrew-Biggs acceleration the code is more than 100x faster on the GPU than on the CPU.
In the case with Andrew-Biggs acceleration the code is only 10x faster on the GPU. I haven't traced what causes the slow-down in the calculations but I assume that some transfer happens between GPU and CPU that is causing the slowdown. It could also be the overhead of python function calls (will have to try putting the acceleration calculations in the main loop).

I have not tried porting the direct convolution method. I think the cuFFT routines are highly optimized. They are also used for the computations in flowdec and pycudadecon.
Further speedups should be possible by re-using the FFT plan as outlined in the issue above.

**Comment by david-hoffman** [comment]
You've convinced me that it's worthwhile. Do you want to wait until I close #4 before submitting a pull request? I'm fine with it either way.

**Comment by VolkerH** [comment]
Hi, thanks for your reply.

Not quite sure to be honest and it will depend on what your aim with #4 is. 

My original modifications were only to your "matlab" core. In the process of trying to figure out why the acceleration computation was much slower for me I made quite a few additional modifications last Friday ... again only for the Matlab  core. This was mainly for my own use and in the current form it may be quite difficult to merge anything back into your code base. It is more of a proof-of-concept. I do think the other RL cores could be similarly converted to cupy but from my limited testing I didn't see any of them offering benefits over the "matlab" core.

**Comment by david-hoffman** [comment]
Proof of concept is fine, I think for now it's best to put it into its own module called `gpu.py` and submit a pull request. We can figure out better integration later.

**Comment by VolkerH** [comment]
Sounds good. I will be on holiday next week. Will try to have something the week after and make a PR.

**Comment by VolkerH** [comment]
With some delay and major changes I now submitted a PR. 

It is not that clean yet and duplicates quite a bit of your existing code in the repo, 
but it is all I found time for at the moment. I removed a few functions (for the RL core and acceleration calculation) and included them into the inner loop, which probably does not help with readability. I did this originally as I noticed a bottleneck somewhere and thought it may be due to function calls (it probably wasn't).
